cinder_plugin/volume.py

```diff
@@ -114,13 +200,91 @@ def get_attachment(cinder_client, volume_id, server_id):
             return attachment
 
 
+def _get_snapshot_name(ctx, kwargs):
+    return "vol-{}-{}".format(get_openstack_id(ctx), kwargs["snapshot_name"])
+
+
+@with_cinder_client
+def snapshot_create(cinder_client, **kwargs):
+    volume_id = get_openstack_id(ctx)
+
+    backup_name = _get_snapshot_name(ctx, kwargs)
+
+    snapshot_incremental = kwargs["snapshot_incremental"]
+    if not snapshot_incremental:
+        ctx.logger.info("Backup create: {}".format(backup_name))
+        cinder_client.backups.create(volume_id, name=backup_name)
+    else:
+        ctx.logger.info("Snapshot create: {}".format(backup_name))
+        description = kwargs.get("snapshot_type", "")
+        cinder_client.volume_snapshots.create(volume_id,
+                                              force=True,
+                                              name=backup_name,
+                                              description=description,
+                                              metadata=None)
+
+
+@with_cinder_client
+def snapshot_apply(cinder_client, **kwargs):
+    volume_id = get_openstack_id(ctx)
+
+    backup_name = _get_snapshot_name(ctx, kwargs)
+    snapshot_incremental = kwargs["snapshot_incremental"]
+    if not snapshot_incremental:
+        ctx.logger.info("Backup apply {} to {}".format(backup_name, volume_id))
+        search_opts = {
+            'volume_id': volume_id,
+            'name': backup_name
+        }
+
+        backups = cinder_client.backups.list(
```

Comment:
Should we be handling multiple backup restores in the same operation?

Reply:
What do you mean?